### PR TITLE
feat(ui5-time-picker): Redesign the Time Picker component according to the new design

### PR DESCRIPTION
**Pull request by NHristov-sap**
The Time Picker component will be redesigned according to the new design with clocks, already provided for UI5 TimePicker control.

Instead of currently existing sliders, new clock dial components whould be used to select time (hours/minutes/seconds, each selected by separate clock). Also, there is AM/PM segmented button that replaces the AM/PM slider.

image

image

In addition to mouse, touch or mouse wheel time selection, there is improved keyboard interaction added - hours/minutes and seconds can be selected by keyboard ([PageUp]/[PageDown] for hours, [Shift]+[PageUp]/[PageDown] for minutes and [Ctrl/Cmd]+[Shift]+[PageUp]/[PageDown] for seconds as well as [A] or [P] for AM/PM selection). Also, direct number typing is accepted.

Accessibility/SR announcements are also covered.

### CODE REVIEW

File: packages/main/src/TimePickerClock.ts [hunk 1]
```diff
@@ -0,0 +1,1000 @@
...
+ * <code>import "@ui5/webcomponents/dist/TimePickerClock.js";</code>
+ *
+ * @constructor
+ * @author SAP SE
+ * @alias sap.ui.webc.main.TimePickerClock
+ * @extends sap.ui.webc.base.UI5Element
+ * @abstract
+ * @tagname ui5-time-picker-clock
+ * @since 1.??.??
+ * @public
```
Comment (ilhan007) [inline packages/main/src/TimePickerClock.ts:77]: New public component ui5-time-picker-clock, is this intended?
I thought we redesign the `ui5-time-picker`

**Review by ilhan007 — CHANGES_REQUESTED** [review]

File: packages/main/src/TimePickerClock.ts (hunk 1 above)
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.ts:77]: Err, sorry, my mistake, it will be private...
In general, it is part of another, also private control, that will replace sliders.

**Review by ilhan007 — DISMISSED** [review]
The PR says:
"refactor(ui5-time-picker): Redesign the Time Picker component according to the new design"

but there is no redesign of the TimePicker, it's a PR that adds private web component (that I assume will be used later with another changed in the TimePicker). The PR description needs adjustments.

File: packages/main/src/TimePickerClock.hbs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+>
+	<div
+		data-label="{{label}}"
+		class="ui5-tp-clock-dial"
+	>
+	</div>
+
+	<div>
+	{{#each _items}}
+		<div class="ui5-tp-clock-item" style="--tpc-degree: {{this.angle}}deg">
```
Comment (ilhan007) [inline packages/main/src/TimePickerClock.hbs:16]: @vladitasev could you check if the following is CSP compliant?
```html
<div class="ui5-tp-clock-item" style="--tpc-degree: {{this.angle}}deg">
```

File: packages/main/src/themes/TimePickerClock.css [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	transform: translate(-50%) rotate(var(--tpc-degree));
+}
+
+.ui5-tp-clock-item .ui5-tp-clock-number {
+	transform: rotate(var(--tpc-degree));
+}
+
+
+
+.sapMTPCInvisible {
```
Comment (ilhan007) [inline packages/main/src/themes/TimePickerClock.css:139]: sapMTPCInvisible should be removed?

File: packages/main/test/pages/TimePickerClock.html [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	<div>
+		<h3>Last Changed Clock - ID, value and string value</h3>
+		<ui5-input id="clockChangeId"></ui5-input>
+		<ui5-input id="clockChangeValue"></ui5-input>
+		<ui5-input id="clockChangeStringValue"></ui5-input>
+	</div>
+
+	<script>
+
+		myHours12.addEventListener('change', displayChange);
```
Comment (ilhan007) [inline packages/main/test/pages/TimePickerClock.html:58]: In test pages we use the noConflict event names - in your case ui5-change

File: packages/main/test/pages/TimePickerClock.html [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		myHours00.addEventListener('change', displayChange);
+		myHours24.addEventListener('change', displayChange);
+		myMinutes.addEventListener('change', displayChange);
+		myMinutes10.addEventListener('change', displayChange);
+
+		function displayChange(event) {
+			var clock = event.target;
```
Comment (ilhan007) [inline packages/main/test/pages/TimePickerClock.html:65]: Some tests should be definitely added

**Comment by NHristov-sap** [comment]
> The PR says: "refactor(ui5-time-picker): Redesign the Time Picker component according to the new design"
> 
> but there is no redesign of the TimePicker, it's a PR that adds private web component (that I assume will be used later with another changed in the TimePicker). The PR description needs adjustments.

I will change "refactor" with "wip" - in fact the component that is currently in this PR cannot (and have no idea to) be used separately. There will be at least two more components (also private and useless outside Time Picker context) included in this PR and all of them then will be used in the refactoring.

BR,
Niki

**Comment by NHristov-sap** [comment]
Hello, the redesign is still in process, that's why I am reopening this PR.
BR,
Nikolay Hristov

File: packages/main/src/ToggleSpinButton.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+})
+
+class ToggleSpinButton extends ToggleButton {
+	/**
+	 * Defines the ARIA valuemin of the component.
+	 *
+	 * @name sap.ui.webc.main.ToggleSpinButton.prototype.valueMin
+	 * @type {Integer}
+	 * @defaultvalue -1
+	 * @public
```
Comment (ilhan007) [inline packages/main/src/ToggleSpinButton.ts:50]: if the component is private, better remove the @ public tags everywhere

**Review by ilhan007 — COMMENTED** [review]
The change is major, please describe it better in the PR's description

File: packages/main/src/TimePickerClock.ts [hunk 7]
```diff
@@ -0,0 +1,1038 @@
...
+	styles: TimePickerClockCss,
+	template: TimePickerClockTemplate,
+})
+
+/**
+ * Fired when a value of clock is changed.
+ *
+ * @event sap.ui.webc.main.TimePickerClock#change
+ * @public
+ * @param { Integer } value The new <code>value</code> of the clock.
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.ts:96]: Done.

File: packages/main/src/TimePickerClock.ts [hunk 8]
```diff
@@ -0,0 +1,1038 @@
...
+		}
+
+		// hover simulation calculations
+		if (isInnerHover || isOuterHover) {
+			this._hoveredValue = this.support2400 && !is24HoursVisible && this._selectedValue === 0 ? 24 : this._selectedValue;
+		} else {
+			this._hoveredValue = -1;
+		}
+
+		if (this._selectedValue === this._getMaxValue() && this.lastItemReplacement !== -1) {
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.ts:692]: No, because as you can see _selectedValue is used in _hoveredValue calculation, and in the commented line _selectedValue is being modified.

File: packages/main/src/TimePickerClock.ts [hunk 9]
```diff
@@ -0,0 +1,1038 @@
...
+		const itemStep = 360 / CLOCK_ANGLE_STEP / values.length;
+
+		// determines step for values display in units
+		if (this.valueStep * itemStep > displayStep) {
+			displayStep = this.valueStep * itemStep;
+		}
+
+		for (i = 1; i <= 60; i++) {
+			valueIndex = i / itemStep - 1;
+			if (i % displayStep !== 0) {
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.ts:512]: Done.

File: packages/main/src/TimePickerClock.ts [hunk 10]
```diff
@@ -0,0 +1,1038 @@
...
+const CLOCK_NUMBER_SELECTED_CLASS = "ui5-tp-clock-selected";
+const CLOCK_MIDDOT_CLASS = "ui5-tp-clock-mid-dot";
+
+/**
+ * @class
+ *
+ * <h3 class="comment-api-title">Overview</h3>
+ *
+ * <code>ui5-time-picker-clock</code> is the single clock item to use inside a <code>ui5-time-picker-clocks</code>.
+ * <code>ui5-time-picker-clocks</code> is one of the internal controls of <code>ui5-time-picker</code>
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.ts:65]: Done.

File: packages/main/src/TimePickerClock.ts [hunk 11]
```diff
@@ -0,0 +1,1038 @@
...
+	 * @param {number} value The value of the clock item
+	 * @returns {number} The real/replaced value
+	 * @private
+	 */
+	_fixReplacementValue(value: number) {
+		let realValue = value;
+		const maxValue = this.itemMax * (this.innerItems ? 2 : 1);
+
+		if (!this.support2400)	{
+			if (realValue === 0) {
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.ts:428]: Nope... Let me explain: 

The second IF does replacement of the value if (realValue === maxValue && this.lastItemReplacement !== -1).
but here we have 2 cases for the fist IF:

- if realValue is 0, it becomes maxValue
- but also realValue can be maxValue initially

so second IF should be checked always, not only as ELSE.
In addition, there cannot be composition of two IFs in one, like:

if ((realValue === maxValue || realValue === 0)  && this.lastItemReplacement !== -1) {
    realValue = this.lastItemReplacement;
}

because this.lastItemReplacement can be -1

So, to summarize, both IFs should be checked

File: packages/main/src/TimePickerClock.ts [hunk 12]
```diff
@@ -0,0 +1,1038 @@
...
+	_onTouchStart(evt: Event) {
+		this._cancelTouchOut = false;
+
+		if (this.disabled || this._mouseOrTouchDown) {
+			return;
+		}
+
+		const x = evt.type === "touchstart" ? (evt as TouchEvent).touches[0].pageX : (evt as MouseEvent).pageX;
+		const y = evt.type === "touchstart" ? (evt as TouchEvent).touches[0].pageY : (evt as MouseEvent).pageY;
+		// console.warn(x, y);
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.ts:915]: Done.

File: packages/main/src/TimeSelectionClocks.ts [hunk 13]
```diff
@@ -0,0 +1,713 @@
...
+			// direct number enter
+			this._exactMatch = undefined;
+			this._resetCooldown(true);
+
+			if (isColon(evt)) {
+				this._kbdBuffer = "";
+				this._resetCooldown(true);
+				this._switchNextClock(true);
+			} else if (this._clocks[this._activeIndex]) {
+				bufferStr = this._kbdBuffer + char;
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimeSelectionClocks.ts:322]: Done.

File: packages/main/src/ToggleSpinButton.ts (hunk 6 above)
Comment (NHristov-sap) [reply packages/main/src/ToggleSpinButton.ts:50]: Done.

File: packages/main/test/pages/TimePickerClock.html (hunk 5 above)
Comment (NHristov-sap) [reply packages/main/test/pages/TimePickerClock.html:68]: Should be addded.

File: packages/main/test/pages/TimePickerClock.html (hunk 4 above)
Comment (NHristov-sap) [reply packages/main/test/pages/TimePickerClock.html]: Yes, you're right.

File: packages/main/src/themes/TimePickerClock.css (hunk 3 above)
Comment (NHristov-sap) [reply packages/main/src/themes/TimePickerClock.css]: Done.

File: packages/main/src/TimeSelectionClocks.ts [hunk 14]
```diff
@@ -0,0 +1,713 @@
...
+		this._getSeparators();
+
+		this._clocks = [];
+		this._buttons = [];
+		this._periods = [];
+
+		if (this._hasHoursComponent) {
+			// add Hours clock
+			this._componentMap.hours = this._clocks.length;
+			this._clocks.push({
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimeSelectionClocks.ts:445]: I prefer to leave as it is now, because using factory function(s) will not lead to any improvements.

File: packages/main/src/TimeSelectionClocks.ts [hunk 15]
```diff
@@ -0,0 +1,713 @@
...
+	_switchNextClock(wrapAround = false) {
+		let activeIndex = this._activeIndex;
+		const startActiveIndex = activeIndex;
+		const clocksCount = this._clocks.length;
+		const activeClock = this._clockComponent(activeIndex);
+
+		do {
+			activeIndex++;
+			if (activeIndex >= clocksCount) {
+				activeIndex = wrapAround ? 0 : clocksCount - 1;
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimeSelectionClocks.ts:604]: In fact we couldn't, as there are 3 cases.

File: packages/main/src/TimeSelectionClocks.ts [hunk 16]
```diff
@@ -0,0 +1,713 @@
...
+	 * @param {event} evt Event object
+	 * @private
+	 */
+	_hoursChange(evt: CustomEvent<TimePickerClockChangeEventDetail>) {
+		let hours = evt.detail.value;
+		const isTwelveHoursFormat = this._hoursConfiguration.isTwelveHoursFormat;
+
+		if (isTwelveHoursFormat) {
+			if (this._period === this.periodsArray[0]) { // AM
+				hours = hours === 12 ? 0 : hours;
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimeSelectionClocks.ts:671]: Again, there are more than 2 cases.

File: packages/main/src/TimePicker.ts [hunk 17]
```diff
@@ -12,20 +12,19 @@ import {
...
  *
  * <h3>Usage</h3>
  * The user can enter a time by:
  * <ul>
- * <li>Using the sliders that are displayed in a popup</li>
+ * <li>Using the clocks that are displayed in a popup</li>
  * <li>Typing it in directly in the input field</li>
  * </ul>
  * <br><br>
- * When the user makes an entry and chooses the enter key, the sliders shows the corresponding time.
```
Comment (DavidMitreski5) [inline packages/main/src/TimePicker.ts:27]: When the user makes an entry and chooses the enter key, the sliders show the corresponding time.

**Review by DavidMitreski5 — CHANGES_REQUESTED** [review]

**Review by unazko — APPROVED** [review]
Tested on different themes and time formats. All seems to be working fine.

**Review by vladitasev — DISMISSED** [review]
When the clock is open and I change the value of the timepicker in the console
$0.value = "15:00:00" f.e., nothing happens. I must close the clock and reopen it for it to update.
I get warnings in the console for texts that don't exist.

File: packages/main/src/TimePickerClock.hbs (hunk 2 above)
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.hbs:16]: Done.

File: packages/main/src/TimePickerClock.hbs [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+		{{#if this.item includeZero=true}}
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.hbs:17]: Done.

File: packages/main/bundle.common.js [hunk 19]
```diff
@@ -80,9 +80,13 @@ import TableCell from "./dist/TableCell.js";
 import TextArea from "./dist/TextArea.js";
 import TimeSelection from "./dist/TimeSelection.js";
 import TimePicker from "./dist/TimePicker.js";
+import TimePickerClock from "./dist/TimePickerClock.js";
```
Comment (ilhan007) [inline packages/main/bundle.common.js:83]: Aren't these modules imported by TimPicker, TimePickerBase classes. Do we need them here imported explicitly?

File: packages/main/src/TimePickerClock.ts [hunk 20]
```diff
@@ -0,0 +1,991 @@
...
+	/**
+	 * If set to <code>true</code>, an inner circle is displayed.
+	 * The first item value of the inner circle will be itemMax + 1
+	 *
+	 * @name sap.ui.webc.main.TimePickerClock.prototype.innerItems
+	 * @type {boolean}
+	 * @defaultvalue false
+	 */
+	@property({ type: Boolean })
+	innerItems!: boolean;
```
Comment (ilhan007) [inline packages/main/src/TimePickerClock.ts:164]: `innerItems` sounds like a slot of items, for a boolean `"show/hideInnerCircle"` is more appropriate

File: packages/main/src/TimePickerClock.ts [hunk 21]
```diff
@@ -0,0 +1,991 @@
...
+
+	/**
+	 * Label of the clock dial - for example, 'Hours', 'Minutes', or 'Seconds'.
```
Comment (ilhan007) [inline packages/main/src/TimePickerClock.ts:167]: If it's 'Hours', 'Minutes', or 'Seconds', better create use enum type with 3 options

File: packages/main/test/specs/TimePickerClock.spec.js [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+		await browser.pause(500);
+		assert.strictEqual(await hours12Disabled.getProperty("selectedValue"), 12, "There is no value change after click on 1st number on disabled clock");
+
+		await enabled.click({ x: 45, y: -110, skipRelease: true });
+		await browser.pause(500);
+		assert.strictEqual(await hours12.getProperty("selectedValue"), 1, "There is proper value set after click on 1st number on enabled clock");
+	});
+
+	it("'active' property", async () => {
+		await browser.url(`test/pages/TimePickerClock.html`);
```
Comment (ilhan007) [inline packages/main/test/specs/TimePickerClock.spec.js:22]: Don't open the page for every test, this makes the test much more slower.
Create the page as such that you don't need to restart everything.
Use more instances in the HTML page if required.

File: packages/main/test/specs/TimeSelectionClocks.spec.js [hunk 23]
```diff
@@ -0,0 +1,361 @@
...
+		assert.notOk(await clocks[2].getProperty("active"), "Third clock is not active");
+
+		await browser.keys(" ");
+		assert.notOk(await clocks[0].getProperty("active"), "First clock is not active");
+		assert.notOk(await clocks[1].getProperty("active"), "Second clock is not active");
+		assert.ok(await clocks[2].getProperty("active"), "Third clock is active");
+	});
+
+	it("switch AM/PM", async () => {
+		await browser.url(`test/pages/TimeSelectionClocks.html`);
```
Comment (ilhan007) [inline packages/main/test/specs/TimeSelectionClocks.spec.js:63]: same

File: packages/main/bundle.common.js (hunk 19 above)
Comment (ilhan007) [reply packages/main/bundle.common.js:83]: I saw at the end that ui5-time-selection-clocks, ui5-time-picker-clock are used standalone for testing purposes, so TimeSelectionClocks  and TimePickerClock are ok to remain.

But check ToggleSpinButton and TimePickerInternals imports

File: packages/main/src/TimePickerClock.ts (hunk 20 above)
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.ts:164]: renamed to showInnerCircle

File: packages/main/bundle.common.js (hunk 19 above)
Comment (NHristov-sap) [reply packages/main/bundle.common.js:83]: Done, removed.

File: packages/main/src/TimePickerClock.ts (hunk 21 above)
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.ts:167]: The ui5-time-picker-clock is private component, these labels are translatable, and are included in the messagebundle, so I think it is not a good idea to be used as enum. Now these labels are provided internally by the TimePickerInternals - there are getters for them, and directly return he corresponding labels from the message bundle.

Also, if in the future the ui5-time-picker-clock is being used for another purpose, another labels can be used, so I prefer to keep the label as string that can be defined by developer.

File: packages/main/test/specs/TimePickerClock.spec.js (hunk 22 above)
Comment (NHristov-sap) [reply packages/main/test/specs/TimePickerClock.spec.js:22]: Done

File: packages/main/test/specs/TimeSelectionClocks.spec.js (hunk 23 above)
Comment (NHristov-sap) [reply packages/main/test/specs/TimeSelectionClocks.spec.js:63]: Done

File: packages/main/test/specs/TimeSelectionClocks.spec.js [hunk 24]
```diff
@@ -0,0 +1,360 @@
+import { assert } from "chai";
+
+describe("Interactions", () => {
+	before(async () => {
+		await browser.url(`test/pages/TimeSelectionClocks.html`);
```
Comment (ilhan007) [inline packages/main/test/specs/TimeSelectionClocks.spec.js:5]: Could you try opening the "TimeSelectionClocks.html" just once in the Interactions section, and not opening the page again in Accessibility and Events. If you need some kind of resetting, you can use other instances if the time selection clock. It just that we noticed reopening of the page slows down test execution and the PR checks.

File: packages/main/test/specs/TimePicker.spec.js [hunk 25]
```diff
@@ -1,6 +1,7 @@
 import { assert } from "chai";
 
 describe("TimePicker general interaction", () => {
+
 	it("input receives value in format pattern depending on the set language", async () => {
 		await browser.url(`test/pages/TimePicker.html?sap-ui-language=bg`);
```
Comment (ilhan007) [inline packages/main/test/specs/TimePicker.spec.js:6]: We will probably refactor the exiting spec files to minimize the loading of the test page, that's why I am commenting the usage of `await browser.url` only in the newly added specs.

File: packages/main/src/TimePickerClock.ts [hunk 26]
```diff
@@ -0,0 +1,991 @@
...
+	 * @type {integer}
+	 * @defaultvalue -1
+	 */
+	@property({ validator: Integer, defaultValue: -1, noAttribute: true })
+	_longTouchId!: number;
+
+	constructor() {
+		super();
+		// attach global event
+		document.addEventListener("mouseup", this._onMouseOutUp.bind(this), false);
```
Comment (ilhan007) [inline packages/main/src/TimePickerClock.ts:378]: It's a good practice to remove the listener
```js
// See an example in the ListItem.js
onEnterDOM() {
	document.addEventListener("mouseup", this.deactivate););
}

onExitDOM() {
	document.removeEventListener("mouseup", this.deactivate);
}
```

File: packages/main/src/TimePickerInternals.ts [hunk 27]
```diff
@@ -0,0 +1,440 @@
...
+	valid: boolean,
+}
+
+/**
+ * @class
+ *
+ * <h3 class="comment-api-title">Overview</h3>
+ *
+ * <code>ui5-time-picker-internals</code> is helper component that contains shared methods used in <code>ui5-time-selection-clocks</code>
+ * and <ui5-time-selection-inputs> components and should not be used separately.
```
Comment (ilhan007) [inline packages/main/src/TimePickerInternals.ts:49]: - what is `ui5-time-selection-inputs`, is it used 
- is `ui5-time-picker-internals` used somewhere or this is a base class

Comment (NHristov-sap) [reply packages/main/src/TimePickerInternals.ts:49]: The mentioned component (ui5-time-selection-inputs) will be developed and added to the timepicker as a next step - it is a part of the redesign. It will affect only the usage on mobile devices, giving a separate popover with numeric inputs, where simple numeric keyboard will be displayed instead of full one. As the big part of the functionality used by ui5-time-selection-clocks will be used in ui5-time-selection-inputs too, it is separated in private ui5-time-picker-internals component.

File: packages/main/src/TimePickerClock.ts (hunk 26 above)
Comment (NHristov-sap) [reply packages/main/src/TimePickerClock.ts:378]: Done.

File: packages/main/test/specs/TimePickerClock.spec.js [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+
+		assert.strictEqual(await numbersInHours24[numbersInHours24.length-1].getText(), "24", "The last number element in clock with prependZero=false and no lastItemReplacement property set is '24'");
+		assert.strictEqual(await numbersInHours00[numbersInHours00.length-1].getText(), "00", "The last number element in clock with prependZero=true and lastItemReplacement=0 is '00'");
+		assert.strictEqual(await numbersInMinutes[numbersInMinutes.length-1].getText(), "0", "The last number element in clock with prependZero=false and lastItemReplacement=0 is '0'");
+	});
+});
+
+describe("Clock item selection", () => {
+	before(async () => {
+		await browser.url(`test/pages/TimePickerClock.html`);
```
[the remark this answers is not in the document]
Comment (NHristov-sap) [reply packages/main/test/specs/TimePickerClock.spec.js:73]: Done.

File: packages/main/test/specs/TimePicker.spec.js (hunk 25 above)
Comment (NHristov-sap) [reply packages/main/test/specs/TimePicker.spec.js:6]: Done.

**Review by vladitasev — CHANGES_REQUESTED** [review]
Some discussions are still in progress with the design team, let's delay the merge of the PR for a bit.

**Review by vladitasev — APPROVED** [review]

**Review by ilhan007 — APPROVED** [review]

**Comment by petyabegovska** [comment]
Part of: #3678
